[VPlan] Implement VPlan-based cost model for VPReduction, VPExtendedReduction and VPMulAccumulateReduction.

ElvisWang123 opened the pull request:
This patch implement the VPlan-based cost model for VPReduction, VPExtendedReduction and VPMulAccumulateReduction. 

With this patch, we can  calculate the reduction cost by the VPlan-based cost  model so remove the  reduction costs in `precomputeCost()`.

Ref: Original instruction based implementation: https://reviews.llvm.org/D93476

ElvisWang123 committed:
[VPlan] Impl VPlan-based pattern match for ExtendedRed and MulAccRed. NFCI

This patch implement the VPlan-based pattern match for extendedReduction
and MulAccReduction. In above reduction patterns, extened instructions
and mul instruction can fold into reduction instruction and the cost is
free.

We add `FoldedRecipes` in the `VPCostContext` to put recipes that can be
folded into other recipes.

ExtendedReductionPatterns:
    reduce(ext(...))
MulAccReductionPatterns:
    reduce.add(mul(...))
    reduce.add(mul(ext(...), ext(...)))
    reduce.add(ext(mul(...)))
    reduce.add(ext(mul(ext(...), ext(...))))

Ref: Original instruction based implementation:
https://reviews.llvm.org/D93476

davemgreen commented:
The way I think I had imagined this working in a vplan-based cost model would be that the vplan nodes more closely mirrored what the back-end would produce. So there would be a vplan recipe for the extended-reduction, which would be created if it was profitable but would otherwise be relatively easy to cost-model.

I'm not sure if that is still the current plan or not.

fhahn commented:
> The way I think I had imagined this working in a vplan-based cost model would be that the vplan nodes more closely mirrored what the back-end would produce. So there would be a vplan recipe for the extended-reduction, which would be created if it was profitable but would otherwise be relatively easy to cost-model.
> 

Yes I think ideally we would model the add-extend/mul-extend operations explicitly in VPlan, especially if matching it in VPlan would require changes to the order in which the cost is computed. Would add-extend/mul-extend be sufficient or would other recipes be needed as well?

ElvisWang123 commented:
I think currently we need mul-extend-reduction and extend-reduction recipe to model these reduction pattern in the VPlan.

Yes, generating new recipe is good but using new recipe to model these patterns would duplicate lots of codes in the `execute()` function since we lack of  middle-end IRs for these patterns. The new recipes would still need to generate  all vector instructions for recipes that has been folding into it.

fhahn commented:
> I think currently we need mul-extend-reduction and extend-reduction recipe to model these reduction pattern in the VPlan.
>
 
@davemgreen do you think those would be enough?

> Yes, generating new recipe is good but using new recipe to model these patterns would duplicate lots of codes in the `execute()` function since we lack of middle-end IRs for these patterns. The new recipes would still need to generate all vector instructions for recipes that has been folding into it.

This might be a case where gradual lowering would help. We could have a more abstract recipe early on which combines mul-extend in a single recipe, facilitating simple cost-computation. Before code-gen, we can replace the recipe with wide recipes for the adds and extends, so there is no need to duplicate codegen for those, similar to how things are sketched for scalar phis in https://github.com/llvm/llvm-project/pull/114305

davemgreen commented:
Hello. I believe the basic patterns are the ones listed in the summary, which are an extended-add-reduction or an extended mla reduction:
```
reduce(ext(...))
reduce.add(mul(...))
reduce.add(mul(ext(...), ext(...)))
```
In the case of MVE, both the ext's will be the same. The add can be done by setting one of the operands to 1. #92418 is similar, but produce a vector instead of a single output value (a dot-product). Dot product has udot, sdot and usdot that do `partialreduce.add(mul(sext, zext))`, but for MVE the extends need to be the same.

There are also other patterns that come up too. The first I believe should be equivalent to `vecreduce(mul(ext, ext))`, providing the ext nodes are the correct types. I don't remember about the second.
```
reduce.add(ext(mul(ext(...), ext(...))))
reduce.add(ext(mul(...)))
```

AArch64 has a stand-alone umull instruction (both for scalar and for vector, although the type sizes differ), that performs a `mul(ext, ext)`. Sometimes it might be better to fold towards `ext(load)` though, depending on the types.

davemgreen commented:
BTW, I believe this patch is currently changing the scores computed for `reduce.add(ext(mul(ext(...), ext(...))))`, in a way that makes it pick a different VF. I wasn't sure if that was because this was better or not, but I think it is picking different type sizes. https://godbolt.org/z/7TYb8heEM should hopefully show the difference? Let me know if it doesn't.

ElvisWang123 commented:
> This might be a case where gradual lowering would help. We could have a more abstract recipe early on which combines mul-extend in a single recipe, facilitating simple cost-computation. Before code-gen, we can replace the recipe with wide recipes for the adds and extends, so there is no need to duplicate codegen for those, similar to how things are sketched for scalar phis in #114305

Thanks for your advice, I'm working on this direction.


> Hello. I believe the basic patterns are the ones listed in the summary, which are an extended-add-reduction or an extended mla reduction:
> 
> ```
> reduce(ext(...))
> reduce.add(mul(...))
> reduce.add(mul(ext(...), ext(...)))
> ```
Thanks, I will only model these three patterns for reduction.


> There are also other patterns that come up too. The first I believe should be equivalent to `vecreduce(mul(ext, ext))`, providing the ext nodes are the correct types. I don't remember about the second.
> 
> ```
> reduce.add(ext(mul(ext(...), ext(...))))
> reduce.add(ext(mul(...)))
> ```
Thanks for caching that. I misunderstood that how many patterns can be folded into mve reduction-like instructions in the original patch.
> AArch64 has a stand-alone umull instruction (both for scalar and for vector, although the type sizes differ), that performs a `mul(ext, ext)`. Sometimes it might be better to fold towards `ext(load)` though, depending on the types.

I think we already model the  instruction cost for `ext(load)` in `VPWidenCastRecipe::computeCost()`. We compute the CastContextHint which depends on the load/store for the `ext` instructions. But I am not quite sure will ARMTTI handle this pattern correctly or not.

In summary, I think we only need two new recipes for reduction - `reduce(ext)` and `reduce.add(mul(<optional>(ext), <optional>(ext)))` ? 

If there is any question, please let me know.

ElvisWang123 committed:
Partially support Extended-reduction.

davemgreen commented:
> Thanks, I will only model these three patterns for reduction.

I believe that `reduce.add(ext(mul(ext(...), ext(...))))` is mathematically equivalent to `reduce.add(mul(ext(...), ext(...)))`, so that sounds good so long as we can match both. I don't remember about the `reduce.add(ext(mul(..., ...)))` form, it doesn't sound like it would be equivalent. If it turns out we do need it, it should hopefully be fine to look into adding it later.

> I think we already model the instruction cost for ext(load) in VPWidenCastRecipe::computeCost(). We compute the CastContextHint which depends on the load/store for the ext instructions. But I am not quite sure will ARMTTI handle this pattern correctly or not.
>
> In summary, I think we only need two new recipes for reduction - reduce(ext) and reduce.add(mul(<optional>(ext), <optional>(ext))) ?

IIRC CCH was added for getting the extension costs (more) correct under Arm/MVE, so it hopefully does OK in the current scheme. We might need to be a little careful about which one of `red-ext + load` vs `red + ext-load` we prefer for each vplan.

ElvisWang123 committed:
Support MulAccRecipe

ElvisWang123 committed:
Fix servel errors and update tests.

We need to update tests since the generated vector IR will be reordered.

ElvisWang123 committed:
Refactors

Using lamda function to early return when pattern matched.
Leave some assertions.

ElvisWang123 committed:
Fix typos and update printing test

ElvisWang123 commented:
Update to recipe based implementation.

Note that the EVL version is not implement yet, so the test case in the RISCV changed.

davemgreen commented:
Hello. Let me upload some tests of examples that produce different results with this now. I think it might be that `add(sext(mul(sext, sext))` can be optimized to `add(zext(mul(sext, sext))` nowadays, or that there are multiple uses.

davemgreen commented:
This might show it picking a different VF now: https://godbolt.org/z/zda3dvcrx

ElvisWang123 commented:
> Hello. Let me upload some tests of examples that produce different results with this now. I think it might be that `add(sext(mul(sext, sext))` can be optimized to `add(zext(mul(sext, sext))` nowadays, or that there are multiple uses.

Thanks for your information.  I will fix the MulAcc pattern match for this pattern.

davemgreen commented:
This is another one that is behaving differently, I think due to subtleties about when one-use checks are beneficial. https://godbolt.org/z/fYW9Y5TqG. There is a third more awkward case with interleave groups I have not looked into much yet.

I will try and make those into test cases to ensure we have the test coverage for them. I will have to check again later if anything else remains that is behaving differently and hitting the assert.

davemgreen commented:
Some tests in ab9178e3e73dc715463e1019ed2cd449dc18bb18.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2653,6 +2657,260 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+/// a chain. This recipe is high level abstract which will generate
+/// VPReductionRecipe and VPWidenCastRecipe before execution. The Operands are
+/// {ChainOp, VecOp, [Condition]}.
+class VPExtendedReductionRecipe : public VPSingleDefRecipe {
```
How much extra code would be needed to add the required info to the existing VPReductionRecipe? If to much, could we at least avoid most duplication in the new recipe by inheriting from VPReductionRecipe?

ElvisWang123 replied:
Inherited from VPReductionRecipe, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2653,6 +2657,260 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  Type *ResultTy;
+  /// Opcode for the extend instruction.
+  Instruction::CastOps ExtOp;
+  CastInst *ExtInstr;
```
At this point, it would be good to avoid adding new references to IR instructions to recipes, Can it be avoided?

ElvisWang123 replied:
Yes we can remove all underlying instruction but that will generate some duplicate vector instructions when there are multiple MulAcc or ExtendedReduction in the same vector loop and share same extend instructions.
We can find some test changed in [234e81e](https://github.com/llvm/llvm-project/pull/113903/commits/234e81ee87184b277b4f4b17ac222737fbf114eb)

ElvisWang123 replied:
Dropping the these instructions will make VPWidenCast recipe without underlying instruction. So it will generate the vector instruction without metadata. 
I am not sure is dropping the metadata here safe or not?

fhahn replied:
I'm not sure what range metadata would be applicable to the extends?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2653,6 +2657,260 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  /// Opcode for the extend instruction.
+  Instruction::CastOps ExtOp;
+  CastInst *ExtInstr;
+  bool IsZExt;
```
Can info be retrieved from ExtOp?

ElvisWang123 replied:
Fixed, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2653,6 +2657,260 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  Instruction::CastOps ExtOp;
+
+  /// reduce.add(ext(mul(ext0(), ext1())))
+  Instruction *MulInstr;
```
Do we need this duplication with the reduction recipe?

ElvisWang123 replied:
You mean the `ExtOp` duplicate from the extended-reduction recipe?  Or MulInstr?
I am not sure what you mean.

fhahn replied:
Not sure if this is still at the correct line, main point was to avoid duplicating things from the parent class. The general point about avoiding references to IR would still apply to the instruction references here and elsewhere

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2653,6 +2657,260 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  Instruction *MulInstr;
+  CastInst *ExtInstr = nullptr;
+  CastInst *Ext0Instr = nullptr;
+  CastInst *Ext1Instr = nullptr;
```
At this point, it would be good to avoid adding new references to IR instructions to recipes, Can it be avoided?

ElvisWang123 replied:
Currently we cannot. We need to add these underlying instructions in the recipes to avoid assertion of legacy cost model and VPlan-based cost model be skipped. 
[Add extra underlying instructions into `SeenInstrs`](https://github.com/llvm/llvm-project/pull/113903/files#diff-da321d454a7246f8ae276bf1db2782bf26b5210b8133cb59e4d7fd45d0905decR7397)

fhahn replied:
Hmm, if they are only used in `planContainsAdditionalSimplifications`, that's not the strongest motivation to keep them if it can be avoided. As per your latest comment, it sounds like the VPlan-based cost model returns better results in some cases and that's the source for differences.

If that's the case, I'd just return true from `planContainsAdditionalSimplifications` if it contains an VMulAddRecipe

ElvisWang123 replied:
Fixed, thanks. 

If we don't add these folded instruction in the `SeenInst` set, it will skip the assertion.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2185,30 +2186,150 @@ InstructionCost VPReductionRecipe::computeCost(ElementCount VF,
+      CostKind);
+
+  // Check if folding ext into ExtendedReduction is profitable.
+  if (MulAccCost.isValid() &&
```
Can we only create the MullAcc recipe if the cost is valid?

ElvisWang123 replied:
I think it is impossible. The validity of instruction cost of MulAcc adn Extended-reuction depends on the `VF`.

fhahn replied:
But each Plan is only valid for a VF range. It is also split the range, if MulAcc and extended reductions are only valid for parts of the range.

fhahn replied:
Thanks, but why do we still need to compute MulAccCost and Extended costs? Can we only create the recipe if MulAcc is profitable?

fhahn reviewed (commented):
Thanks for the update, some  initial suggestions inline.

Would be good if we could avoid any references to IR operands, and possibly unify both recipes (and avoid duplication with the existing reduction recipe)

ElvisWang123 committed:
Fold reduce.add(zext(mul(sext(A), sext(B)))) into MulAccRecipe when A == B

For the future refactor of avoiding reference underlying instructions
and mismatched opcode and the entend instruction in the new added
pattern, removed passing UI when creating VPWidenCastRecipe.
This removed will lead to dupicate extend instruction created after loop
vectorizer when there are two reduction patterns exist in the same loop.
This redundant instruction might be removed after LV.

ElvisWang123 commented:
> Some tests in [ab9178e](https://github.com/llvm/llvm-project/commit/ab9178e3e73dc715463e1019ed2cd449dc18bb18).

Thanks for adding new tests.

Update the condition of creating MulAccRecipe for `reduce.add(zext(mul(sext(A), sext(B))))` when `A == B`.  The new patch can pass all assertion checks.

To support `mla_and_add_together_16_64` which contains two reduction patterns in same loop, `reduce.add(zext(mul(sext(A), sext(A))))` and `reduce(sext(A))`.  I removed the checks of if the extend recipes has only one user to align the behavior of the legacy model. This change might lead to miscalculating cost  if the extend recipes will be used by other recipes not in the reduction pattern, which we still need to calculate the costs of extend recipes. But I think we would have same  issue in the legacy cost model that we can fix it in future.

ElvisWang123 committed:
Refactor! Reuse functions from VPReductionRecipe.

ElvisWang123 committed:
Refactor! Add comments and refine new recipes.

davemgreen commented:
It makes sense to try and keep these cost models in-sync, but this might be doing better than the legacy one, I am still seeing a fair number of asserts. I've added two new tests to the same file, one has two reductions which have a shared sext node which in the old cost model would be cost-modelled but needn't have (multiple uses were difficult for it to deal with optimally). The other has a `ext(mul(ext, ext))` reduction that isn't handled by the old cost model properly. #115847 should hopefully help bring that case closer to this.

davemgreen commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2653,6 +2657,210 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  Instruction *getMulInstr() const { return MulInstr; }
+
+  /// The underlying Instruction for outer VPWidenCastRecipe.
+  CastInst *getExtInstr() const { return ExtInstr; }
```
Is it worth representing a `reduce(ext(mul(ext, ext)))` reduction as `reduce(mul(ext, ext))`, as (so long as the extends are large enough) they should be equivalent?

ElvisWang123 replied:
Thanks for your comments. 

Without `outer` extend, I can clean up lots of bad naming and refine the `prepareExecute()`.

davemgreen commented:
There might be some other cases with low-trip-count loops too, where we still hit the assert. Is it worth considering keeping the assert disabled for these loops? I have verified that the performance looks OK, often better now and we don't want a worse designed cost model just to appease an overly strict assert.

ElvisWang123 commented:
For these new test cases ([ab9178e](https://github.com/llvm/llvm-project/commit/ab9178e3e73dc715463e1019ed2cd449dc18bb18), [7665d3f](https://github.com/llvm/llvm-project/commit/7665d3f0df78b8b58c1adb18d53f36351426da72)), I would like to raise  discussion of the difference between the legacy model and the VPlan-based cost model for in-loop reduction cost.

1. Should we limit all `ext` and `mul` instructions has only one user in both extended-reduction and MulAcc patterns ?
    -  If Yes, limit all `ext` and `mul` has only one user and the user is reduction.  Bad for VPlan-based cost model but this is how legacy cost model be implemented.
       - VPlan: We cannot fold the `VPWidenCastRecipe` and `VPWidenRecipe` into `VPExtendedReductionRecipe` and `VPMulAccRecipe` when there are two reductions share same `ext` or `mul`. This will generate larger cost than the legacy model and also far from the  assembly.
       - Legacy: Currently the legacy cost model limit these instruction has only one user. This make legacy cost model calculate the cost for `ext` and `mul` even these instruction can be  folded into following reductions.
    -  If No. Good for Vplan based but miscalculate the cost for legacy model.
       - VPlan: Let `VPWidenCastRecipe` and `VPWidenRecipe` can transform into `VPExtendedReductionRecipe` and `VPMulAccRecipe`. If the `WidenCast` and `Widen` recipes has other user than reductions, it will still calculate the cost of `WidenCast` and `Widen` recipes. This seems much closer to the generated assembly.
       -  Legacy: If `ext` and `mul` instructions has users other than reductions, we will not calculate the cost of these instructions.
     - Current status: 
     Currently I removed some `OneUser` limitations in the VPlan-bsed implementation to support two reductions in one loop and the legacy cost model always limit `ext` and `mul` instructions has one user. But in the new test case (`test_std_q31`), the legacy cost model cannot mark the `ext` instruction to free because it is shared by two reductions. This mis-calculation from the legacy model trigger the assertion and I think it is difficult to fix it in the legacy model.
2. Do we want this patch to land before assertion is removed?

fhahn reviewed (commented):
If the legacy-cost estimates are worse, I don't think we need to try hard to match them, if we can make use of more accurate modeling in VPlan. If that's the case, I'd use `planContainsAdditionalSimplifications` to allow divergences between legacy and VPlan-based cost model for plans that contain extended reduction recipes.

davemgreen commented:
> If the legacy-cost estimates are worse, I don't think we need to try hard to match them, if we can make use of more accurate modeling in VPlan. If that's the case, I'd use `planContainsAdditionalSimplifications` to allow divergences between legacy and VPlan-based cost model for plans that contain extended reduction recipes.

Yeah I think that makes sense in this regard, if we can avoid the assert. The old cost model found it difficult to deal with things that the new method can more easily handle, and I've tried to verify the performance looks OK with the new costmodel.

ElvisWang123 committed:
Remove underying instruction dependency.

ElvisWang123 committed:
Revert "Remove underying instruction dependency."

This reverts commit e4e510802f44c9936f5a892eb16ba2b19651ee15.

We need underlying instruction for accurate TTI query and
the metadata from to attach on the generating vector instructions.

ElvisWang123 committed:
Remove extended instruction after mul in MulAccRecipe.

Note that reduce.add(ext(mul(ext(A), ext(A)))) is mathmetical equal
to reduce.add(mul(zext(A), zext(A))).

ElvisWang123 committed:
Refactor.

- Move tryToMatch{MulAcc|ExtendedReduction} out from
adjustRecipesForReduction().
- Remove `ResultTy` which is same as the recurrence type.
- Use VP_CLASSOF_IMPL.

ElvisWang123 committed:
Clamp the range when the ExtendedReduction or MulAcc cost is invalid.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9419,9 +9389,81 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
-      VPReductionRecipe *RedRecipe =
-          new VPReductionRecipe(RdxDesc, CurrentLinkI, PreviousLink, VecOp,
-                                CondOp, CM.useOrderedReductions(RdxDesc));
+      auto TryToMatchMulAcc = [&]() -> VPReductionRecipe * {
```
Move out from the  adjustRecipesForReductions(). Thanks.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2571,6 +2573,8 @@ class VPReductionRecipe : public VPSingleDefRecipe {
   }
 
+  // TODO: Support VPExtendedReductionRecipe and VPMulAccRecipe after EVL
+  // support.
```
Drop it. Thanks.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2653,6 +2657,210 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  static inline bool classof(const VPUser *U) {
+    auto *R = dyn_cast<VPRecipeBase>(U);
+    return R && classof(R);
+  }
```
Fixed, thanks.

ElvisWang123 committed:
Try to not depend on underlying ext/mul instructions and preserve flags/DL.

This commit expand the ctor of VPWidenRecipe and the VPWidenCastRecipe
to generate vector instruction with debug information and flags without
underlying instruction.

The extend instruction (ZExt) will need non-negative flags. and the int mul
instructions will need no-unsigned-wrap/no-signed-wrap flags.

ElvisWang123 committed:
Update testcase and fix reduction cost.

TTI should model the cost of move result of reduction to the scalar
register and add/sub/cmp... to the scalar value. We don't need to add
the cost of binOp in the vplan-based cost model.

ElvisWang123 commented:
Rebase and update test cases.

ElvisWang123 commented:
Gentle ping!

ElvisWang123 committed:
!fixup. Rebase to upstream `prepareToExecute()` implementation.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2192,30 +2193,146 @@ InstructionCost VPReductionRecipe::computeCost(ElementCount VF,
+  TTI::TargetCostKind CostKind = TTI::TCK_RecipThroughput;
+  unsigned Opcode = RdxDesc.getOpcode();
+
+  // ExtendedReduction Cost
```
comment as is is redundant

ElvisWang123 replied:
Removed, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2192,30 +2193,146 @@ InstructionCost VPReductionRecipe::computeCost(ElementCount VF,
+                                       VPCostContext &Ctx) const {
+  const RecurrenceDescriptor &RdxDesc = getRecurrenceDescriptor();
+  RecurKind RdxKind = RdxDesc.getRecurrenceKind();
+  Type *ElementTy = getResultType();
```
can we instead update inferScalarType to do the correct thing for the recipe?

ElvisWang123 replied:
Updated, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2192,30 +2193,146 @@ InstructionCost VPReductionRecipe::computeCost(ElementCount VF,
+InstructionCost VPMulAccRecipe::computeCost(ElementCount VF,
+                                            VPCostContext &Ctx) const {
+  const RecurrenceDescriptor &RdxDesc = getRecurrenceDescriptor();
+  Type *ElementTy = isExtended() ? RdxDesc.getRecurrenceType()
+                                 : Ctx.Types.inferScalarType(getVecOp0());
```
If possible inferScalarType should compute the correct type. Enough to get the result type?

ElvisWang123 replied:
Updated, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2192,30 +2193,146 @@ InstructionCost VPReductionRecipe::computeCost(ElementCount VF,
+
+  assert(Opcode == Instruction::Add &&
+         "Reduction opcode must be add in the VPMulAccRecipe.");
+  // MulAccReduction Cost
```
Redundant comment

ElvisWang123 replied:
Removed, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2192,30 +2193,146 @@ InstructionCost VPReductionRecipe::computeCost(ElementCount VF,
+  InstructionCost ReductionCost = Ctx.TTI.getArithmeticReductionCost(
+      Opcode, VectorTy, RdxDesc.getFastMathFlags(), CostKind);
+
+  // Extended cost
```
Redundant comment

ElvisWang123 replied:
Removed, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1820,12 +1820,100 @@ void VPlanTransforms::createInterleaveGroups(
            vp_depth_first_deep(Plan.getEntry()))) {
-    for (VPRecipeBase &R : make_early_inc_range(VPBB->phis())) {
+    for (VPRecipeBase &R : make_early_inc_range(*VPBB)) {
+      if (auto *ExtRed = dyn_cast<VPExtendedReductionRecipe>(&R)) {
+        expandVPExtendedReduction(ExtRed);
+        continue;
+      }
+      if (auto *MulAcc = dyn_cast<VPMulAccRecipe>(&R)) {
+        expandVPMulAcc(MulAcc);
+        continue;
+      }
```
Better to handle phis first and then take care of other recipes?

ElvisWang123 replied:
Updated thanks.
Handle phis first then handle other recipes.

ElvisWang123 committed:
Move VPReductionRecipe inherite from VPRecipeWithIRFlags.

This commit make 2 major changes.

1. VPReductionRecipe inherite form VPRecipeWithIRFlags.
2. VPMulacc/VPExtendedReduction also inherite from VPRecipeWithIRFlags.

Note that we need to disable droping flags for VPReductionRecipe in
`clearReductionWithIRFlags()` to prevent flags of Mul dropped.

Note that we temporary prevent the EVL transformation for
VPExtendedReductionRecipe and VPMulAccRecipe since we have not
implemeted the EVL-version of these recipes.

This is mostly NFC only few vplan-printing change.

ElvisWang123 committed:
Only create VPMulAcc/VPExtendedReduction recipe when beneficial. NFC

Only create the VPMulaccRecipe and the VPExtendedReductionRecipe when it's beneficial.

Also address comments and cleanup computeCost().

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2694,6 +2724,202 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  Instruction::CastOps ExtOp;
+  DebugLoc ExtDL;
+  /// Non-negative flag for the extended instruction.
+  bool IsNonNeg;
```
Updated, thanks.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2694,6 +2724,202 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
 
+/// A recipe to represent inloop extended reduction operations, performing a
+/// reduction on a vector operand into a scalar value, and adding the result to
+/// a chain. This recipe is high level abstract which will generate
```
Fixed, thanks.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2694,6 +2724,202 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  bool IsNonNeg;
+
+protected:
+  VPExtendedReductionRecipe(const unsigned char SC,
+                            const RecurrenceDescriptor &R, Instruction *RedI,
+                            Instruction::CastOps ExtOp, DebugLoc ExtDL,
+                            bool IsNonNeg, VPValue *ChainOp, VPValue *VecOp,
+                            VPValue *CondOp, bool IsOrdered)
```
I think it's protected. For future EVL-version recipes.

fhahn replied:
Ok, can this be introduce when needed?

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

Resolve conflicts and prevent VPReductionRecipe::print() crach since
not all VPReductionRecipe contains underlying instruction.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9387,6 +9330,171 @@ VPlanPtr LoopVectorizationPlanner::buildVPlan(VFRange &Range) {
+                                             Type *SrcTy) -> bool {
+    return LoopVectorizationPlanner::getDecisionAndClampRange(
+        [&](ElementCount VF) {
+          VectorType *SrcVecTy = cast<VectorType>(ToVectorTy(SrcTy, VF));
+          VectorType *VectorTy = cast<VectorType>(ToVectorTy(RedTy, VF));
```
```suggestion
          auto *SrcVecTy = cast<VectorType>(ToVectorTy(SrcTy, VF));
          auto *VectorTy = cast<VectorType>(ToVectorTy(RedTy, VF));
```

ElvisWang123 replied:
Fixed, thanks.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9387,6 +9330,171 @@ VPlanPtr LoopVectorizationPlanner::buildVPlan(VFRange &Range) {
+          TTI::TargetCostKind CostKind = TTI::TCK_RecipThroughput;
+          Type *SrcTy =
+              Ext0 ? Ctx.Types.inferScalarType(Ext0->getOperand(0)) : RedTy;
+          VectorType *SrcVecTy = cast<VectorType>(ToVectorTy(SrcTy, VF));
+          VectorType *VectorTy = cast<VectorType>(ToVectorTy(RedTy, VF));
```
```suggestion
          auto *SrcVecTy = cast<VectorType>(ToVectorTy(SrcTy, VF));
          auto *VectorTy = cast<VectorType>(ToVectorTy(RedTy, VF));
```

ElvisWang123 replied:
Fixed, thanks.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9387,6 +9330,171 @@ VPlanPtr LoopVectorizationPlanner::buildVPlan(VFRange &Range) {
+
+  // Try to match reduce.add(mul(...))
+  if (match(VecOp, m_Mul(m_VPValue(A), m_VPValue(B)))) {
+    VPWidenCastRecipe *RecipeA =
+        dyn_cast_if_present<VPWidenCastRecipe>(A->getDefiningRecipe());
+    VPWidenCastRecipe *RecipeB =
+        dyn_cast_if_present<VPWidenCastRecipe>(B->getDefiningRecipe());
+    VPWidenRecipe *Mul = cast<VPWidenRecipe>(VecOp->getDefiningRecipe());
```
```suggestion
    auto *RecipeA =
        dyn_cast_if_present<VPWidenCastRecipe>(A->getDefiningRecipe());
    auto *RecipeB =
        dyn_cast_if_present<VPWidenCastRecipe>(B->getDefiningRecipe());
    auto *Mul = cast<VPWidenRecipe>(VecOp->getDefiningRecipe());
```

ElvisWang123 replied:
Fixed, thanks.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9387,6 +9330,171 @@ VPlanPtr LoopVectorizationPlanner::buildVPlan(VFRange &Range) {
+    VPWidenRecipe *Mul = cast<VPWidenRecipe>(VecOp->getDefiningRecipe());
+
+    // Matched reduce.add(mul(ext, ext))
+    if (RecipeA && RecipeB && match(RecipeA, m_ZExtOrSExt(m_VPValue())) &&
+        match(RecipeB, m_ZExtOrSExt(m_VPValue())) &&
+        (RecipeA->getOpcode() == RecipeB->getOpcode() || A == B)) {
```
```suggestion
    if (RecipeA && RecipeB &&
       (RecipeA->getOpcode() == RecipeB->getOpcode() || A == B) &&
        match(RecipeA, m_ZExtOrSExt(m_VPValue())) &&
        match(RecipeB, m_ZExtOrSExt(m_VPValue()))) {
```

ElvisWang123 replied:
Reordered, thanks.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9387,6 +9330,171 @@ VPlanPtr LoopVectorizationPlanner::buildVPlan(VFRange &Range) {
+    // which can be transform to reduce.add(zext(mul(sext(A), sext(B)))).
+  } else if (match(VecOp, m_ZExtOrSExt(m_Mul(m_ZExtOrSExt(m_VPValue()),
+                                             m_ZExtOrSExt(m_VPValue()))))) {
+    VPWidenCastRecipe *Ext =
+        cast<VPWidenCastRecipe>(VecOp->getDefiningRecipe());
+    VPWidenRecipe *Mul =
+        cast<VPWidenRecipe>(Ext->getOperand(0)->getDefiningRecipe());
+    VPWidenCastRecipe *Ext0 =
+        cast<VPWidenCastRecipe>(Mul->getOperand(0)->getDefiningRecipe());
+    VPWidenCastRecipe *Ext1 =
+        cast<VPWidenCastRecipe>(Mul->getOperand(1)->getDefiningRecipe());
```
```suggestion
    auto *Ext =
        cast<VPWidenCastRecipe>(VecOp->getDefiningRecipe());
    auto *Mul =
        cast<VPWidenRecipe>(Ext->getOperand(0)->getDefiningRecipe());
    auto *Ext0 =
        cast<VPWidenCastRecipe>(Mul->getOperand(0)->getDefiningRecipe());
    auto *Ext1 =
        cast<VPWidenCastRecipe>(Mul->getOperand(1)->getDefiningRecipe());
```

ElvisWang123 replied:
Fixed, thanks.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2196,30 +2197,49 @@ InstructionCost VPReductionRecipe::computeCost(ElementCount VF,
+InstructionCost VPMulAccRecipe::computeCost(ElementCount VF,
+                                            VPCostContext &Ctx) const {
+  Type *RedTy = Ctx.Types.inferScalarType(this);
+  VectorType *SrcVecTy =
+      cast<VectorType>(ToVectorTy(Ctx.Types.inferScalarType(getVecOp0()), VF));
```
```suggestion
  auto *SrcVecTy =
      cast<VectorType>(ToVectorTy(Ctx.Types.inferScalarType(getVecOp0()), VF));
```

ElvisWang123 replied:
Fixed, thanks.

ElvisWang123 committed:
!fixup use `auto`

ElvisWang123 committed:
!fixup VPReductionRecipe unit tests.

Update unit tests to contruct VPReductionRecipe without null underlying
instruction. With null instruction, it will fail to in the ctor of
VPRecipeWithIRFlags.

Note that this commit will drop the underlying instruction when transform
VPReductionRecipe to VPReductionEVLRecipe but preserve the debug
location.

davemgreen commented:
The Arm tests still look OK to me, it looks like this could do with a rebase/test fix.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9387,6 +9330,170 @@ VPlanPtr LoopVectorizationPlanner::buildVPlan(VFRange &Range) {
+          InstructionCost RedCost = Ctx.TTI.getArithmeticReductionCost(
+              Instruction::Add, VectorTy, RdxDesc.getFastMathFlags(), CostKind);
+          InstructionCost ExtCost = 0;
+          if (Ext0)
+            ExtCost += Ext0->computeCost(VF, Ctx);
+          if (Ext1)
+            ExtCost += Ext1->computeCost(VF, Ctx);
+          if (OuterExt)
+            ExtCost += OuterExt->computeCost(VF, Ctx);
+
+          return MulAccCost.isValid() &&
+                 MulAccCost < ExtCost + MulCost + RedCost;
```
Great to see that we can make this decision here without the legacy cost model!

This should also allow us to move this to VPlanTransforms.cpp?

ElvisWang123 replied:
Migrated, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2579,7 +2611,7 @@ class VPInterleaveRecipe : public VPRecipeBase {
 /// a vector operand into a scalar value, and adding the result to a chain.
 /// The Operands are {ChainOp, VecOp, [Condition]}.
-class VPReductionRecipe : public VPSingleDefRecipe {
+class VPReductionRecipe : public VPRecipeWithIRFlags {
```
Why change this?

ElvisWang123 replied:
As your [suggestion](https://github.com/llvm/llvm-project/pull/113903#discussion_r1875680827), the flags of extended/MulAcc reduction should be handled by `VPRecipeWithFlags`.   So the flags can be dropped when needed.
Meanwhile,  extend/MulAcc reduction recipes share lots of function with `VPReductionRecipe`.

But inherit from both `VPRecipeWIthIRFlags` and `VPReductionRecipe` for VPExtended/MulAcc recipes  will lead to multiple inheritance (diamond) of VPRecipeBase. Calling function in `VPRecipeBase` will be ambiguous.

Perhaps we can split off the change of VPReductionReductionRecipe and the unit-test update? 
Or choose  a single class to inherit to prevent multiple inheritance?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2694,6 +2772,201 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+/// [Condition]}.
+class VPExtendedReductionRecipe : public VPReductionRecipe {
+  Instruction::CastOps ExtOp;
+  DebugLoc RedDL;
```
Document please why we need multiple debug locations.

ElvisWang123 replied:
Updated, thanks.

ElvisWang123 replied:
Removed, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2694,6 +2772,201 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  ~VPExtendedReductionRecipe() override = default;
+
+  VPExtendedReductionRecipe *clone() override {
+    llvm_unreachable("Not implement yet");
```
This is because it will be replaced by concrete recipes before cloning if needed for epilogue vectorization?

ElvisWang123 replied:
Implemented, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2694,6 +2772,201 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+/// multiplication into a scalar value, and adding the result to a chain. This
+/// recipe is abstract and needs to be lowered to concrete recipes before
+/// codegen. The Operands are {ChainOp, VecOp1, VecOp2, [Condition]}.
+class VPMulAccRecipe : public VPReductionRecipe {
```
Would be good to include Reduction in the name here, for consistency, how about `VPMulAccumulateReductionRecipe`?

ElvisWang123 replied:
Updated, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2694,6 +2772,201 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+class VPMulAccRecipe : public VPReductionRecipe {
+  Instruction::CastOps ExtOp;
+
+  // Non-neg flag for the extend instruction.
+  bool IsNonNeg;
+
+  // Debug location of extend instruction.
+  DebugLoc Ext0DL;
+  DebugLoc Ext1DL;
+
+  // Debug location of reduction instruction.
+  DebugLoc RedDL;
+
+  // Is this mul-acc recipe contains extend recipes?
+  bool IsExtended = false;
```
Use `///` for doc-comments

ElvisWang123 replied:
Fixed, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1504,49 +1504,50 @@ void VPWidenCastRecipe::execute(VPTransformState &State) {
     setFlags(CastOp);
 }
 
+// Computes the CastContextHint from a recipes that may access memory.
```
```suggestion
/// Computes the CastContextHint for a recipe.
```

It computes the win for any recipe I think

ElvisWang123 replied:
Updated, thanks.

ElvisWang123 committed:
!fixup migrate tryTo* to VPlanTransforms

ElvisWang123 committed:
Implement clone() and add some docs.

ElvisWang123 committed:
Update comments.

ElvisWang123 committed:
fix-ReductionEVLRecipe query underlyingInstr().

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

ElvisWang123 committed:
Update after merge.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9387,6 +9330,170 @@ VPlanPtr LoopVectorizationPlanner::buildVPlan(VFRange &Range) {
+static VPMulAccRecipe *tryToMatchAndCreateMulAcc(
+    const RecurrenceDescriptor &RdxDesc, Instruction *CurrentLinkI,
+    VPValue *PreviousLink, VPValue *VecOp, VPValue *CondOp,
+    LoopVectorizationCostModel &CM, VPCostContext &Ctx, VFRange &Range) {
```
Migrated, thanks.

ElvisWang123 commented on llvm/test/Transforms/LoopVectorize/RISCV/inloop-reduction.ll:
```diff
@@ -187,26 +187,36 @@ define i32 @add_i16_i32(ptr nocapture readonly %x, i32 %n) {
+; IF-EVL-INLOOP-NEXT:    [[TMP17:%.*]] = select <vscale x 8 x i1> [[TMP9]], <vscale x 8 x i32> [[TMP12]], <vscale x 8 x i32> zeroinitializer
+; IF-EVL-INLOOP-NEXT:    [[TMP14:%.*]] = call i32 @llvm.vector.reduce.add.nxv8i32(<vscale x 8 x i32> [[TMP17]])
+; IF-EVL-INLOOP-NEXT:    [[TMP11]] = add i32 [[TMP14]], [[VEC_PHI]]
+; IF-EVL-INLOOP-NEXT:    [[INDEX_EVL_NEXT]] = add nuw i32 [[TMP5]], [[EVL_BASED_IV]]
```
Prevent create these abstract recipes when using tail-folding with EVL since we haven't supported yet.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9611,9 +9555,30 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
+      if (ForceTailFoldingStyle == TailFoldingStyle::DataWithEVL) {
+        RedRecipe = new VPReductionRecipe(RdxDesc, CurrentLinkI, PreviousLink,
+                                          VecOp, CondOp, IsOrderedRed);
+      } else {
+        if (auto *MulAcc =
```
```suggestion
      } else if (auto *MulAcc =
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2613,6 +2688,13 @@ class VPReductionRecipe : public VPSingleDefRecipe {
+                    VPValue *VecOp, VPValue *CondOp, bool IsOrdered,
+                    DebugLoc DL = {})
+      : VPReductionRecipe(VPDef::VPReductionSC, R,
+                          ArrayRef<VPValue *>({ChainOp, VecOp}), CondOp,
```
```suggestion
                          {ChainOp, VecOp}, CondOp,
```
?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2702,6 +2785,212 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+                            VPValue *ChainOp, VPWidenCastRecipe *Ext,
+                            VPValue *CondOp, bool IsOrdered)
+      : VPReductionRecipe(VPDef::VPExtendedReductionSC, R,
+                          ArrayRef<VPValue *>({ChainOp, Ext->getOperand(0)}),
```
```suggestion
                          {ChainOp, Ext->getOperand(0)},
```
??

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2702,6 +2785,212 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  Instruction::CastOps ExtOp;
+
+  /// Non-neg flag of the extend recipe.
+  bool IsNonNeg;
```
```suggestion
  bool IsNonNeg = false;
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1523,49 +1523,50 @@ void VPWidenCastRecipe::execute(VPTransformState &State) {
+/// Computes the CastContextHint for a recipe.
+static TTI::CastContextHint computeCCH(const VPRecipeBase *R, ElementCount VF) {
+  if (VF.isScalar())
+    return TTI::CastContextHint::Normal;
```
```suggestion
    return TTI::CastContextHint::None;
```
?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1851,6 +1859,88 @@ void VPlanTransforms::createInterleaveGroups(
+    // VPWidenCastRecipe.
+    if (MulAcc->getVecOp0() == MulAcc->getVecOp1()) {
+      Op1 = Op0;
+    } else {
+      if (MulAcc->isZExt())
```
```suggestion
    } else if (MulAcc->isZExt())
```

fhahn replied:
still pending?

ElvisWang123 replied:
There exists a `else` for the `if` and come of the codes shared between them.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1851,6 +1859,88 @@ void VPlanTransforms::createInterleaveGroups(
+  }
+
+  // Generate VPWidenRecipe.
+  SmallVector<VPValue *, 2> MulOps = {Op0, Op1};
```
std::array?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1867,9 +1957,173 @@ void VPlanTransforms::convertToConcreteRecipes(VPlan &Plan) {
       PhiR->eraseFromParent();
     }
+    for (VPRecipeBase &R : make_early_inc_range(*VPBB)) {
+      if (!isa<VPExtendedReductionRecipe, VPMulAccumulateReductionRecipe>(&R))
+        continue;
```
This check can be dropped

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1867,9 +1957,173 @@ void VPlanTransforms::convertToConcreteRecipes(VPlan &Plan) {
+    for (VPRecipeBase &R : make_early_inc_range(*VPBB)) {
+      if (!isa<VPExtendedReductionRecipe, VPMulAccumulateReductionRecipe>(&R))
+        continue;
+      if (auto *ExtRed = dyn_cast<VPExtendedReductionRecipe>(&R)) {
+        expandVPExtendedReduction(ExtRed);
+      }
+      if (auto *MulAcc = dyn_cast<VPMulAccumulateReductionRecipe>(&R)) {
+        expandVPMulAccumulateReduction(MulAcc);
+      }
```
Drop extra braces, one-line substatement

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1867,9 +1957,173 @@ void VPlanTransforms::convertToConcreteRecipes(VPlan &Plan) {
+      return new VPMulAccumulateReductionRecipe(
+          RdxDesc, CurrentLinkI, PreviousLink, CondOp, IsOrderedRed, Mul,
+          RecipeA, RecipeB);
+    } else {
```
```suggestion
    }
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1867,9 +1957,173 @@ void VPlanTransforms::convertToConcreteRecipes(VPlan &Plan) {
+      return new VPMulAccumulateReductionRecipe(
+          RdxDesc, CurrentLinkI, PreviousLink, CondOp, IsOrderedRed, Mul);
+    }
+  } else if (match(VecOp, m_ZExtOrSExt(m_Mul(m_ZExtOrSExt(m_VPValue()),
```
```suggestion
  }
  if (match(VecOp, m_ZExtOrSExt(m_Mul(m_ZExtOrSExt(m_VPValue()),
```

ElvisWang123 committed:
Address comments and split off abstract recipes creation from adjustRecipesForReduction.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -233,7 +233,7 @@ static cl::opt<PreferPredicateTy::Option> PreferPredicateOverEpilogue(
                          "tail-folding fails.")));
 
-static cl::opt<TailFoldingStyle> ForceTailFoldingStyle(
+cl::opt<TailFoldingStyle> ForceTailFoldingStyle(
```
I don't think this is the right approach to check ForceTailFoldingStyle, better to use `CM.foldTailWithEVL()`

ElvisWang123 replied:
Fixed, thanks.

ElvisWang123 committed:
!fixup using foldTailWithEVL.

And address comments.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1928,3 +2024,174 @@ void VPlanTransforms::handleUncountableEarlyExit(
+                                               VFRange &Range) {
+  // TODO: Remove EVL check when we support EVL version of
+  // VPExtendedReductionRecipe and VPMulAccumulateReductionRecipe.
+  if (Ctx.foldTailWithEVL())
```
Better to not call the transform instead of checking here?

ElvisWang123 replied:
Updated, thanks!

ElvisWang123 committed:
!fixup, remove extra debugLoc and move check of EVL out of transforms.

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

ElvisWang123 committed:
Update after merge main.

* Use Ctx.CostKind.
* ToVectorTy -> toVectorTy.
* Add bc.resume.val into scalar.ph

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

Resolve conflicts and update removed headers.

ElvisWang123 commented:
Ping :)

ElvisWang123 committed:
Update after merge. Using runPass::().

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1263,10 +1264,17 @@ void VPlanTransforms::clearReductionWrapFlags(VPlan &Plan) {
 
-    for (VPUser *U : collectUsersRecursively(PhiR))
+    for (VPUser *U : collectUsersRecursively(PhiR)) {
+      // Flags in reduction recipes are control/store by the recurrence
+      // descriptor. Dropping flags for VPExtendedReductionRecipe and
+      // VPMulaccRecipe may drop flags that we don't expect to drop.
+      if (isa<VPReductionRecipe>(U))
```
What case is this covering? If it is using VPRecipeWithIRFlags to manage flags, they should be dropped here as needed I think?

ElvisWang123 replied:
Removed, thanks!

ElvisWang123 committed:
!fixup, Remove unused check and functions.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1204,10 +1234,17 @@ class VPWidenCastRecipe : public VPRecipeWithIRFlags {
 
-  VPWidenCastRecipe(Instruction::CastOps Opcode, VPValue *Op, Type *ResultTy)
-      : VPRecipeWithIRFlags(VPDef::VPWidenCastSC, Op), Opcode(Opcode),
+  VPWidenCastRecipe(Instruction::CastOps Opcode, VPValue *Op, Type *ResultTy,
+                    DebugLoc DL = {})
+      : VPRecipeWithIRFlags(VPDef::VPWidenCastSC, Op, DL), Opcode(Opcode),
         ResultTy(ResultTy) {}
 
+  VPWidenCastRecipe(Instruction::CastOps Opcode, VPValue *Op, Type *ResultTy,
+                    bool IsNonNeg, DebugLoc DL = {})
+      : VPRecipeWithIRFlags(VPDef::VPWidenCastSC, Op, NonNegFlagsTy(IsNonNeg),
+                            DL),
+        Opcode(Opcode), ResultTy(ResultTy) {}
+
```
It could be split off but it's hard to test it through input IR with `nneg` flags.

Currently all the VPWidenCastRecipes  contains `nneg` flags are directly constructed from the scalar IR.
When constructing VPWidenCastRecipe with input IR, `VPRecipeWithIRFlags` will automatically assign the corresponding flags. 

But when expanding VPMulAccReductionRecipe and VPExtendedReductionRecipes, we already drop the underlying IR for the cast instructions. So we need to pass `nneg` flags to the VPWIdenCastRecipe.

ElvisWang123 commented:
Gental ping :)

davemgreen commented:
Could you rebase? I gave this a try and it still seemed to do OK on MVE. I wanted to check other archs though as I was seeing some odd errors to do with the cost model being different and I wanted to check that they were spurious.

NFCI can be removed from the title now.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2329,8 +2366,45 @@ class VPReductionRecipe : public VPSingleDefRecipe {
-      : VPSingleDefRecipe(SC, Operands, I, DL), RdxDesc(R),
+                    VPValue *CondOp, bool IsOrdered)
+      : VPRecipeWithIRFlags(SC, Operands, *I), RdxDesc(R),
+        IsOrdered(IsOrdered) {
+    if (CondOp) {
+      IsConditional = true;
```
```suggestion
        IsOrdered(IsOrdered), IsCondition(CondOp) {
    if (CondOp) {
```
?

ElvisWang123 replied:
Fixed, thanks.

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

ElvisWang123 committed:
!fixup; Address comments.

ElvisWang123 commented:
Hi @davemgreen 
I've updated to latest main. Please try again.

davemgreen commented:
Does this fail for you under aarch64? I didn't look into what might be going on - it was strange that it was trying to do any vectorization with fp128 long doubles.
https://godbolt.org/z/hP4eYooTK

ElvisWang123 commented:
> Does this fail for you under aarch64? I didn't look into what might be going on - it was strange that it was trying to do any vectorization with fp128 long doubles. https://godbolt.org/z/hP4eYooTK

Hi @davemgreen 
I am not sure I understood correctly.
Did you mean that the cost model should return InvalidCost or the loop vectorizer should not try to vectorize loop entirely when the type is fp128?

I indeed found an assertion failure on this case because the cost of `fmuladd` from vplan-based cost model  misaligned to the legacy one. Will open other PR or update this PR to fix the cost of `fmuladd` in vplan-based cost model.

davemgreen commented:
> > Does this fail for you under aarch64? I didn't look into what might be going on - it was strange that it was trying to do any vectorization with fp128 long doubles. https://godbolt.org/z/hP4eYooTK
> 
> Hi @davemgreen I am not sure I understood correctly. Did you mean that the cost model should return InvalidCost or the loop vectorizer should not try to vectorize loop entirely when the type is fp128?

I just meant that I don't think there is any point in vectorizing fp128 - there is no native support so it will all become libcalls anyway and there isn't much point in "vectorizing" (or interleaving with a vf==1). The cost should maybe be higher than they are but not infinite. It is hopefully not too important though if the error is coming from the fmuladd on a double, and this code just happens to trigger the mismatch in costs.

ElvisWang123 committed:
!fixup, Add Mul cost to prevent FMuladd Reduction cost misaligned.

davemgreen commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2173,6 +2261,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
   LatchExitingBranch->eraseFromParent();
 }
 
+/// This function try to match following pattern to create
```
try->tries

davemgreen commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2173,6 +2261,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
+/// This function try to match following pattern to create
+/// VPExtendedReductionRecipe and clamp the \p Range if it is beneficial and
+/// valid. The created VPExtendedReductionRecipe will lower to concrete before
+/// executeion.
```
execution

davemgreen commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2173,6 +2261,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
 
+/// This function try to match following pattern to create
+/// VPExtendedReductionRecipe and clamp the \p Range if it is beneficial and
+/// valid. The created VPExtendedReductionRecipe will lower to concrete before
```
to concrete recipes?

davemgreen reviewed (commented):
The test now all pass just fine and the MVE improvements are still present. Thanks for the fix.

ElvisWang123 committed:
!Fixup, typo.

fhahn reviewed (commented):
Thanks for the latest updates and @davemgreen for checking.

A few more comments from my side, then I think we should be good to go!

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

Resolve conflicts and fix the correctness issue.

ElvisWang123 committed:
!fixup, Address comments and fix VPReductionRecipe::computeCost

Note that we should use std::nullopt when quering cost of non-FPMathOperator instructions.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2313,30 +2319,50 @@ InstructionCost VPReductionRecipe::computeCost(ElementCount VF,
+      cast<VectorType>(toVectorTy(Ctx.Types.inferScalarType(getVecOp()), VF));
+
+  return Ctx.TTI.getExtendedReductionCost(Opcode, isZExt(), RedTy, SrcVecTy,
+                                          RdxDesc.getFastMathFlags(),
```
We should use `getFastMathFlags()` here. 
This should be `getExtendedReductionCost(Opcode, isZExt(), RetTy, SecVecTy, std::nullopt,...)`which is similar to `getArithmeticReductionCost(...)` for non-FP reduction . 
Currently the `getExtendedReductionCost()` doesn't support an optional parameter, I will a separate PR to fix the TTI issue.

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

ElvisWang123 committed:
!fixup, Update after merge, using std::array.

After 6dba5f659, we can pass nullopt that prevent wrong cost from the extended
reduction.

Also remove the ArrayRef since the elements will be used after free.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2073,6 +2074,87 @@ void VPlanTransforms::createInterleaveGroups(
+  }
+
+  // Generate VPWidenRecipe.
+  std::array<VPValue *, 2> MulOps = {Op0, Op1};
```
No. `Op0` and `Op1` will be free before usage.

ElvisWang123 committed:
fixup, formatting.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2185,6 +2273,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
   LatchExitingBranch->eraseFromParent();
 }
 
+/// This function tries to match following pattern to create
+/// VPExtendedReductionRecipe and clamp the \p Range if it is beneficial and
+/// valid. The created VPExtendedReductionRecipe will lower to concrete recipes
+/// before execution.
+///   reduce(ext(...)).
```
```suggestion
/// This function tries convert extended in-loop reductions to VPExtendedReductionRecipe and clamp the \p Range if it is beneficial and
/// valid. The created VPExtendedReductionRecipe must be lower to concrete recipes before execution.
```

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2185,6 +2273,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
+
+  Type *RedTy = Ctx.Types.inferScalarType(Red);
+  VPValue *VecOp = Red->getVecOp();
+  const RecurrenceDescriptor &RdxDesc = Red->getRecurrenceDescriptor();
```
Does this need anything from RdxDesc other than the info retained in @lukel97 's https://github.com/llvm/llvm-project/pull/131300

lukel97 replied:
Yeah, the recurrence kind can be obtained from just the opcode so I think this is fine

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2185,6 +2273,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
+          auto *SrcVecTy = cast<VectorType>(toVectorTy(SrcTy, VF));
+          TTI::TargetCostKind CostKind = TTI::TCK_RecipThroughput;
+          InstructionCost ExtRedCost = Ctx.TTI.getExtendedReductionCost(
+              Opcode, isZExt, RedTy, SrcVecTy, RdxDesc.getFastMathFlags(),
```
Should use fast math flags from recipe?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2185,6 +2273,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
+void VPlanTransforms::convertToAbstractRecipes(VPlan &Plan, VPCostContext &Ctx,
+                                               VFRange &Range) {
+  for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(
+           vp_depth_first_deep(Plan.getEntry()))) {
```
enought to start at entry of vector-loop-region to start with?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.h:
```diff
@@ -178,6 +179,13 @@ struct VPlanTransforms {
   /// Lower abstract recipes to concrete ones, that can be codegen'd.
   static void convertToConcreteRecipes(VPlan &Plan);
 
+  /// This function convert initial recipes to the abstract recipes and clamp \p
```
```suggestion
  /// This function converts initial recipes to the abstract recipes and clamps \p
```

fhahn commented on llvm/test/Transforms/LoopVectorize/vplan-printing.ll:
```diff
@@ -1200,6 +1200,254 @@ exit:
+; CHECK-NEXT: Successor(s): middle.block
+; CHECK-EMPTY:
+; CHECK-NEXT: middle.block:
+; CHECK-NEXT:   EMIT vp<%8> = compute-reduction-result ir<%r.09>, vp<%6>
+; CHECK-NEXT:   EMIT vp<%9> = extract-from-end vp<%8>, ir<1>
+; CHECK-NEXT:   EMIT vp<%cmp.n> = icmp eq ir<%n>, vp<%1>
+; CHECK-NEXT:   EMIT branch-on-cond vp<%cmp.n>
+; CHECK-NEXT: Successor(s): ir-bb<for.cond.cleanup.loopexit>, scalar.ph
+; CHECK-EMPTY:
+; CHECK-NEXT: scalar.ph:
+; CHECK-NEXT:   EMIT vp<%bc.resume.val> = resume-phi vp<%1>, ir<0>
+; CHECK-NEXT:   EMIT vp<%bc.merge.rdx> = resume-phi vp<%8>, ir<0>
+; CHECK-NEXT: Successor(s): ir-bb<for.body>
+; CHECK-EMPTY:
+; CHECK-NEXT: ir-bb<for.body>:
+; CHECK-NEXT:   IR   %i.010 = phi i32 [ %inc, %for.body ], [ 0, %for.body.preheader ] (extra operand: vp<%bc.resume.val> from scalar.ph)
+; CHECK-NEXT:   IR   %r.09 = phi i64 [ %add, %for.body ], [ 0, %for.body.preheader ] (extra operand: vp<%bc.merge.rdx> from scalar.ph)
+; CHECK-NEXT:   IR   %arrayidx = getelementptr inbounds i16, ptr %x, i32 %i.010
+; CHECK-NEXT:   IR   %load0 = load i16, ptr %arrayidx, align 4
+; CHECK-NEXT:   IR   %arrayidx1 = getelementptr inbounds i16, ptr %y, i32 %i.010
+; CHECK-NEXT:   IR   %load1 = load i16, ptr %arrayidx1, align 4
+; CHECK-NEXT:   IR   %conv0 = sext i16 %load0 to i32
+; CHECK-NEXT:   IR   %conv1 = sext i16 %load1 to i32
+; CHECK-NEXT:   IR   %mul = mul nsw i32 %conv0, %conv1
+; CHECK-NEXT:   IR   %conv = sext i32 %mul to i64
+; CHECK-NEXT:   IR   %add = add nsw i64 %r.09, %conv
+; CHECK-NEXT:   IR   %inc = add nuw nsw i32 %i.010, 1
+; CHECK-NEXT:   IR   %exitcond = icmp eq i32 %inc, %n
+; CHECK-NEXT: No successors
+; CHECK-EMPTY:
+; CHECK-NEXT: ir-bb<for.cond.cleanup.loopexit>:
+; CHECK-NEXT:   IR   %add.lcssa = phi i64 [ %add, %for.body ] (extra operand: vp<%9> from middle.block)
+; CHECK-NEXT: No successors
```
I think you could remove those check lines here, they aren't really relevant.

fhahn commented on llvm/test/Transforms/LoopVectorize/vplan-printing.ll:
```diff
@@ -1200,6 +1200,254 @@ exit:
+; CHECK-NEXT: }
+;
+entry:
+  %cmp8 = icmp sgt i32 %n, 0
```
this shouldnt be needed I think

fhahn commented on llvm/test/Transforms/LoopVectorize/vplan-printing.ll:
```diff
@@ -1200,6 +1200,254 @@ exit:
+  %cmp8 = icmp sgt i32 %n, 0
+  br i1 %cmp8, label %for.body, label %for.cond.cleanup
+
+for.body:                                         ; preds = %entry, %for.body
+  %i.010 = phi i32 [ %inc, %for.body ], [ 0, %entry ]
+  %r.09 = phi i64 [ %add, %for.body ], [ 0, %entry ]
```
nit: suggested re-naming for consistency with other tests and slightly better readability.

```suggestion
loop:
  %iv = phi i32[ 0, %entry ],  [ %iv.next, %for.body ], 
  %rdx = phi i64 [ 0, %entry ], [ %tdx.next, %for.body ], 
```

fhahn commented on llvm/test/Transforms/LoopVectorize/vplan-printing.ll:
```diff
@@ -1200,6 +1200,254 @@ exit:
+  %exitcond = icmp eq i32 %inc, %n
+  br i1 %exitcond, label %for.cond.cleanup, label %for.body
+
+for.cond.cleanup:                                 ; preds = %for.body, %entry
```
```suggestion
exit: 
```

nit: slightly more compact.

ElvisWang123 committed:
!fixup, address comments.

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2361,6 +2423,164 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+
+  /// The scalar type after extended.
+  Type *getResultType() const {
+    return getRecurrenceDescriptor().getRecurrenceType();
```
Actually I think if #131300 lands then you'll need to store the recurrence type separately. But that doesn't seem too bad.

fhahn replied:
That should be fine and also clearer if stored in the recipe.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9549,6 +9488,21 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
+  // clamp the range for better cost estimation.
+  // TODO: Enable following transform when the EVL-version of extended-reduction
+  // and mulacc-reduction are implemented.
+  if (!CM.foldTailWithEVL()) {
```
Why do we need to special case this for EVL? Shouldn't the cost-model tell us that the combined reductions aren't profitable?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2262,6 +2294,28 @@ class VPReductionRecipe : public VPRecipeWithIRFlags {
   }
 
+  /// For VPExtendedReductionRecipe.
+  /// Note that IsNonNeg flag and the debug location are for extend instruction.
```
```suggestion
  /// Note that IsNonNeg flag and the debug location are from the extend.
```

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2376,30 +2382,59 @@ InstructionCost VPReductionRecipe::computeCost(ElementCount VF,
+  if (ElementTy->isFloatingPointTy())
+    return Ctx.TTI.getArithmeticReductionCost(Opcode, VectorTy, FMFs,
+                                              Ctx.CostKind);
+  // Cannot get correct cost when quering TTI with FMFs not contains `reassoc`
+  // for non-FP reductions.
```
Ah that's interesting. In that case, I think it would be clearer to pull out FMFs into an optional variable and have a single call to `Ctx.TTI.getArithmeticReductionCost(` if all other operands are the same

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2376,30 +2382,55 @@ InstructionCost VPReductionRecipe::computeCost(ElementCount VF,
+  Type *RedTy = Ctx.Types.inferScalarType(this);
+  auto *SrcVecTy =
+      cast<VectorType>(toVectorTy(Ctx.Types.inferScalarType(getVecOp0()), VF));
+
```
```suggestion
```

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2445,6 +2476,60 @@ void VPReductionEVLRecipe::print(raw_ostream &O, const Twine &Indent,
+    getCondOp()->printAsOperand(O, SlotTracker);
+  }
+  O << ")";
+  if (RdxDesc.IntermediateStore)
+    O << " (with final reduction value stored in invariant address sank "
+         "outside of loop)";
```
This can be removed I think, as the store is explicit in the VPlan anyways nw

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2445,6 +2476,60 @@ void VPReductionEVLRecipe::print(raw_ostream &O, const Twine &Indent,
+    getCondOp()->printAsOperand(O, SlotTracker);
+  }
+  O << ")";
+  if (RdxDesc.IntermediateStore)
+    O << " (with final reduction value stored in invariant address sank "
+         "outside of loop)";
```
Can be removed I think

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.h:
```diff
@@ -14,6 +14,7 @@
 #define LLVM_TRANSFORMS_VECTORIZE_VPLANTRANSFORMS_H
 
 #include "VPlan.h"
+#include "VPlanHelpers.h"
```
can forward declare `VPCostContext` instead of including `VPlanHelpers.h`?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2262,6 +2294,28 @@ class VPReductionRecipe : public VPRecipeWithIRFlags {
+  }
+
+  /// For VPMulAccumulateReductionRecipe.
+  /// Note that the NUW/NSW and DL are for mul instruction.
```
```suggestion
  /// Note that the NUW/NSW and DL are from the Mul instruction.
```

fhahn reviewed (commented):
Thanks for all the updates. A few more comments, then I think it should be good to go.

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

ElvisWang123 committed:
!fixup, formatting and address comments.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9515,6 +9458,25 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
+  // TODO: Use a better method that only set the VF for plan once.
+  Plan->clearVF();
+  for (ElementCount VF : Range)
+    Plan->addVF(VF);
```
Is there a better method to update the VF of the plan?

`handelUncountbleEarlyExit()` will check if the VF of the plan only contains scalarVF. So need to set the VF before it.
But the Range will update after convertToAbstractRecipes. Need to update the VF of plan by Range.

I tried to move the `handleUncountableEarlyExit()`  after `convertToAbstractRecipes` but cause more errors.

fhahn replied:
Ah that's an unfortunate interaction. Let me see if we can break this up somehow. Maybe there's also a better way to handle the `handleUncountableEarlyExit` case.

In the meantime, can we remove the restricted VFs instead of clearing and re-adding?

ElvisWang123 replied:
Updated, thanks.

fhahn replied:
This can be removed now?

ElvisWang123 replied:
No, I just move addVF to plan after `convertToAbstractRecipe()`. Don't need to update the VF anymore.
https://github.com/llvm/llvm-project/pull/113903/files#diff-da321d454a7246f8ae276bf1db2782bf26b5210b8133cb59e4d7fd45d0905decL9760-L9763

fhahn replied:
Ah yes, looks like I looked at an earlier version of the diff, which still had clearing VFs and also TODO.

ElvisWang123 replied:
I think moving this is a NFC. And without this patch we don't need to move this.

Since we will further limit the VF range in `convertToAbstractRecipe`, we need to set the VF for plan after that. And `convertToAbstractRecipe` need to after `adjustRecipesForReductions` since this patch will try to patter match reduction recipe.

sdesmalen-arm replied:
Okay, I see what you mean. Because Range is updated, we need to move this after `convertToAbstractRecipes`, so this is not a non-functional change.

ElvisWang123 commented:
Ping :)

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2356,6 +2419,178 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+#endif
+
+  /// The scalar type after extended.
+  Type *getResultType() const { return ResultTy; }
```
Do we need to update also VPTypeAnalysis to handle the new recipes?

ElvisWang123 committed:
!fixup, Update inferScalarType and not clear the VF of plan.

sdesmalen-arm commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -596,13 +598,15 @@ class VPRecipeWithIRFlags : public VPSingleDefRecipe {
   };
 
+  struct NonNegFlagsTy {
+    char NonNeg : 1;
```
nit: (to keep with convention in this class).
```suggestion
    char IsNonNeg : 1;
```

Also, can this just be a `bool` rather than a bitfield struct? (I don't expect any other NonNegFlags to be added)

sdesmalen-arm commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2189,6 +2277,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
+      return new VPMulAccumulateReductionRecipe(Red, Mul);
+  }
+  // Matched reduce.add(ext(mul(ext(A), ext(B))))
+  // All extend recipes must have same opcode or A == B
```
Is that a limitation that can be relaxed in the future? I presume the answer is 'yes', but just want to check there isn't some fundamental assumption that makes it hard to change this.

We'd need that for e.g. `sudot` instructions, which implements `reduce.add(mul(sext(A), `ext(B))`.

sdesmalen-arm reviewed (commented):
Thanks for this work @ElvisWang123, this is also very useful for SVE instructions. The overall design of this looks good to me, but I've left a few minor comments.

ElvisWang123 committed:
!fixup, address comments.

sdesmalen-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9515,6 +9458,29 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
+                             CostCtx, Range);
+  }
+
+  // Update VF after `convertToAbstractRecipes`. Cannot set the VF here since
+  // `handleUncountableEarlyExit` will check the VF of the plan, need to set
+  // before it and update.
+  // TODO: Use a better method that only set the VF for plan once.
+  SmallVector<ElementCount, 2> VFToRemove;
+  for (ElementCount VF : Plan->vectorFactors())
+    if (!Range.contains(VF))
+      VFToRemove.push_back(VF);
+
+  for (ElementCount VF : VFToRemove)
+    Plan->removeVF(VF);
```
Thanks. I'll resolve this comment when you've landed + rebased and removed the code above.

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

* Resolve conflicts.
* Set VF after `convertToAbstractRecipe()`.
* Move printing tests into vplan-printing-reductions.ll.

ElvisWang123 committed:
!fixup, address comments.

* Fix typos.
* Remove unused functions.

fhahn reviewed (commented):
Thanks for the latest updates! Could you

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -811,6 +823,8 @@ class VPRecipeWithIRFlags : public VPSingleDefRecipe {
 
   FastMathFlags getFastMathFlags() const;
 
+  bool isNonNeg() const { return NonNegFlags.NonNeg; }
```
Can we assert that OpType == OperationType::NonNegOp here, like for other accessors? otherwise we would access a non-active union element.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2341,6 +2373,28 @@ class VPReductionRecipe : public VPRecipeWithIRFlags {
   }
 
+  /// For VPExtendedReductionRecipe.
+  /// Note that IsNonNeg flag and the debug location are from the extend.
```
```suggestion
  /// Note that the IsNonNeg flag and the debug location are from the extend.
```

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2341,6 +2373,28 @@ class VPReductionRecipe : public VPRecipeWithIRFlags {
+  }
+
+  /// For VPMulAccumulateReductionRecipe.
+  /// Note that the NUW/NSW and DL are from the Mul.
```
```suggestion
  /// Note that the NUW/NSW flags and the debug location are from the Mul.
```

(for consistency with comment above)

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2439,6 +2502,175 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+/// A recipe to represent inloop extended reduction operations, performing a
+/// reduction on a extended vector operand into a scalar value, and adding the
+/// result to a chain. This recipe is abstract and needs to be lowered to
+/// concrete recipes before codegen. The Operands are {ChainOp, VecOp,
```
```suggestion
/// concrete recipes before codegen. The operands are {ChainOp, VecOp,
```

ElvisWang123 replied:
Oops. I missed that. Updated, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2439,6 +2502,175 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+             VPSlotTracker &SlotTracker) const override;
+#endif
+
+  /// The scalar type after extended.
```
```suggestion
  /// The scalar type after extending.
```

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2439,6 +2502,175 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  /// Is the extend ZExt?
+  bool isZExt() const { return getExtOpcode() == Instruction::ZExt; }
+
+  /// The Opcode of extend recipe.
```
```suggestion
  /// The opcode of extend recipe.
```

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2439,6 +2502,175 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+/// reduction.add on the result of vector operands (might be extended)
+/// multiplication into a scalar value, and adding the result to a chain. This
+/// recipe is abstract and needs to be lowered to concrete recipes before
+/// codegen. The Operands are {ChainOp, VecOp1, VecOp2, [Condition]}.
```
```suggestion
/// codegen. The operands are {ChainOp, VecOp1, VecOp2, [Condition]}.
```

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2439,6 +2502,175 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  VPValue *getVecOp0() const { return getOperand(1); }
+  VPValue *getVecOp1() const { return getOperand(2); }
+
+  /// Return if this MulAcc recipe contains extend instructions.
```
more accurate? Would be good to strip reference to `instructions`.
```suggestion
  /// Return if this MulAcc recipe contains extended operands.
```

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2439,6 +2502,175 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  /// Return the opcode of the underlying extend.
+  Instruction::CastOps getExtOpcode() const { return ExtOp; }
+
+  /// Return if the extend opcode is ZExt.
```
```suggestion
  /// Return if the operands are zero extended.
```

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2439,6 +2502,175 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+  /// Return if this MulAcc recipe contains extend instructions.
+  bool isExtended() const { return ExtOp != Instruction::CastOps::CastOpsEnd; }
+
+  /// Return the opcode of the underlying extend.
```
```suggestion
  /// Return the opcode of the extends for the operands.
```

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:
```diff
@@ -272,6 +272,14 @@ Type *VPTypeAnalysis::inferScalarType(const VPValue *V) {
             // TODO: Use info from interleave group.
             return V->getUnderlyingValue()->getType();
           })
+          .Case<VPExtendedReductionRecipe>(
+              [](const VPExtendedReductionRecipe *R) {
+                return R->getResultType();
+              })
+          .Case<VPMulAccumulateReductionRecipe>(
+              [](const VPMulAccumulateReductionRecipe *R) {
+                return R->getResultType();
+              })
```
I think something like this should work:

```suggestion
          .Case<VPExtendedReductionRecipe, VPMulAccumulateReductionRecipe>(
              [](const VPExtendedReductionRecipe *R) {
                return R->getResultType();
              })
```

SamTebbs33 replied:
The only common super class for them is `VPReductionRecipe`, so the lambda parameter would need to be cast. In which case it's probably cleaner to keep the structure as it is.

fhahn replied:
The lambda should have used `const auto *R` (updated the suggestion) which is something that should be supported by the type switch, I think we use that feature in a number of places

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanHelpers.h:
```diff
@@ -78,6 +78,11 @@ struct VFRange {
     return End.getKnownMinValue() <= Start.getKnownMinValue();
   }
 
+  bool contains(const ElementCount &VF) const {
+    return VF.getKnownMinValue() >= Start.getKnownMinValue() &&
+           VF.getKnownMinValue() < End.getKnownMinValue();
+  }
+
```
Still needed in the latest version?

SamTebbs33 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2439,6 +2502,175 @@ class VPReductionEVLRecipe : public VPReductionRecipe {
+    return Copy;
+  }
+
+  VP_CLASSOF_IMPL(VPDef::VPMulAccumulateReductionSC);
```
Since building on top of this I've noticed that this class needs a case in `VPRecipeBase::mayHaveSideEffects`, probably above the `VPReductionSC` case.

ElvisWang123 committed:
!fixup, address comments and using `transferFlags()` to copy nneg.

Only VPWidenCastRecipe with ZExt continas nneg flag. Using
transferFlags() to prevent get the nneg flag from a not-nneg flags
operands.

sdesmalen-arm commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -706,6 +710,12 @@ class VPRecipeWithIRFlags : public VPSingleDefRecipe {
 
+  template <typename IterT>
+  VPRecipeWithIRFlags(const unsigned char SC, IterT Operands,
+                      NonNegFlagsTy NonNegFlags, DebugLoc DL = {})
```
nit: 
I'm wondering if it makes sense to make `NonNegFlags` a `std::optional<NonNegFlagsTy>`, and initialise it with `{}` by default (here and in the other classes), so that you don't need to do:
```
if (Something->isZExt())
  ... = new VPWidenCastRecipe(..., Something->isNonNeg());
else
  ... = new VPWidenCastRecipe(...);
```
and instead just do:
```
std::optional<NonNegFlagsTy> NonNegFlags = Something->isZExt() ? Something->isNonNeg() : {};
... = new VPWidenCastRecipe(..., NonNegFlags);
```
What do you think?

sdesmalen-arm replied:
You're right, ignore my suggestion, this isn't as easy as I thought it was.

sdesmalen-arm reviewed (changes requested):
A few more minor comments.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2548,6 +2630,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
+  Type *RedTy = Ctx.Types.inferScalarType(Red);
+  VPValue *VecOp = Red->getVecOp();
+
+  // Test if using extended-reduction is profitable and clamp the range.
```
Done!

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2548,6 +2630,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
+  Type *RedTy = Ctx.Types.inferScalarType(Red);
+
+  // Test if using mulutiply-accumulate-reduction is beneficial and clamp the
+  // range.
```
Update, thanks.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.h:
```diff
@@ -181,6 +182,13 @@ struct VPlanTransforms {
+  /// This function converts initial recipes to the abstract recipes and clamps
+  /// \p Range based on cost model for following optimizations and cost
+  /// estimations. The converted abstract recipes will lower to concrete
+  /// recipies before codegen.
```
Done!

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2548,6 +2630,172 @@ void VPlanTransforms::handleUncountableEarlyExit(
+                                               VFRange &Range) {
+  for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(
+           vp_depth_first_deep(Plan.getVectorLoopRegion()))) {
+    for (VPRecipeBase &R : make_early_inc_range(*VPBB)) {
```
Updated, thanks!

ElvisWang123 committed:
!fixup, address comments.

sdesmalen-arm reviewed (approved):
Thanks for addressing the comments @ElvisWang123!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -782,19 +784,25 @@ Value *VPInstruction::generate(VPTransformState &State) {
+
+    Type *ResTy = Ctx.Types.inferScalarType(this);
+    if (!vputils::onlyFirstLaneUsed(this))
+      ResTy = toVectorTy(ResTy, VF);
+
     if (!getUnderlyingValue()) {
-      // TODO: Compute cost for VPInstructions without underlying values once
-      // the legacy cost model has been retired.
-      return 0;
+      switch (getOpcode()) {
+      case Instruction::FMul:
+        return Ctx.TTI.getArithmeticInstrCost(getOpcode(), ResTy, Ctx.CostKind);
+      default:
+        // TODO: Compute cost for VPInstructions without underlying values once
+        // the legacy cost model has been retired.
+        return 0;
+      }
```
Are those changes here still needed in the latest version? I don't think we create FMul VPInstructions now?

ElvisWang123 replied:
Yes. Otherwise it will hit the assertion of `legacy cost model and vplan-based cost model mismatch` when the reduction is `fmuladd`.

Fmuladd reduction will generate `Instruction::FMul` + fadd reduction. Without the cost of fmul, the cost will slightly lower than the legcy model.

Could be split of in different patch but cannot show the failure without removing `precompute` reduction costs.

The potential failure: 
(C) https://godbolt.org/z/hP4eYooTK 
(llvm IR) https://godbolt.org/z/G9bT4Y1c9

fhahn replied:
I see, is it possible we are missing tests for that? Removing the code seems to not cause any test failures.

Could you add tests showing the difference?

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -155,6 +155,8 @@ bool VPRecipeBase::mayHaveSideEffects() const {
   case VPBlendSC:
   case VPReductionEVLSC:
   case VPReductionSC:
+  case VPExtendedReductionSC:
```
Added, thanks!

ElvisWang123 committed:
!fixup, Add new recipes to mayReadWriteMemory.

fhahn reviewed (commented):
Thanks for all the updates. Basically looks good, but I think it would be good to split up the patch into adding the new recipes + transform + lowering but without the cost changes. And then land the cost changes separately, if possible?

fhahn commented:
That way there will less churn in case we need to revert due to cost failures

ElvisWang123 commented:
Split new recipes implementations to #137745.
Split transformations (convertTo{Abstract|Concrete}Recipes) to #137746.

Will update this patch after above patches landed.
This patch will focus on removing the dependency of the legacy cost model and enable this transformation by VPlan-based cost model.

fhahn commented:
> Split new recipes implementations to #137745. Split transformations (convertTo{Abstract|Concrete}Recipes) to #137746.
> 

Would be good to combine  #137745 & #137746, so the recipes are used already, sorry if that wasn't clearer from the previous comment

ElvisWang123 commented:
> Would be good to combine #137745 & #137746, so the recipes are used already, sorry if that wasn't clearer from the previous comment

Ah it is fine, I misunderstood the previous comment :smiley:. Combined all the changes into #137746.

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

ElvisWang123 committed:
Fixup after merge.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2527,24 +2533,47 @@ InstructionCost VPReductionRecipe::computeCost(ElementCount VF,
+  // and the BinOp cost in the getReductionCost().
   if (RecurrenceDescriptor::isMinMaxRecurrenceKind(RdxKind)) {
     Intrinsic::ID Id = getMinMaxReductionIntrinsicOp(RdxKind);
-    return Cost +
-           Ctx.TTI.getMinMaxReductionCost(Id, VectorTy, FMFs, Ctx.CostKind);
```
Remove the BinOp cost here to match the legacy cost model.  [link](https://github.com/llvm/llvm-project/blob/ec7a5091655f75839e8da57a76aec6600a864915/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp#L5888)
And TTI already calculate the cost of BinOp (at least in RISCV). [link](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp#L1884-L1920)

ElvisWang123 reviewed (commented):
Update after #137746.

sdesmalen-arm reviewed (approved):
Thanks for rebasing @ElvisWang123. Given that this PR is now mostly about flipping the switch to the new cost model, I see no issues with landing it.

fhahn reviewed (approved):
LGTM, thanks

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform

ElvisWang123 committed:
Address comments.

ElvisWang123 committed:
Merge branch 'main' into vp-arm-mve-transform
